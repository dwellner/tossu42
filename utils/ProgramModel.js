import DateUtils from "./DateUtils";
import ProgramService from "../data/ProgramService";

const getIntensityFactor = runType =>
  ({
    iv: 2.5,
    m: 2,
    r: 2,
    t: 2,
    k: 1.5,
    pi: 1.5,
    pe: 1.5,
    ve: 0.5
  }[runType] || 1);

const getIVTotalDistance = (ivDistance, repeats) => {
  const warmup = 2,
    coolDown = 1,
    recovery = 0.4;
  return warmup + (ivDistance + recovery) * repeats + coolDown;
};

const getDayDistance = day => {
  if (day.type === "iv") return getIVTotalDistance(day.distance, day.repeat);
  if (day.type === "lepo") return 0;
  return day.distance;
};

const getDayIntensity = day =>
  getIntensityFactor(day.type) * getDayDistance(day);

const getWeekDistance = week =>
  week.map(getDayDistance).reduce((a, b) => a + b);

const getIntensity = week => week.map(getDayIntensity).reduce((a, b) => a + b);

const getStretchedProgramStart = (program, programLength, totalLength) => {
  let repeats = [];

  const stretchLengthInWeeks = Math.ceil(totalLength - programLength);
  const repeatBlock = program.weeks.slice(0, program.stretchRules.repeat);
  const numberOfRepeatBlocks = Math.ceil(
    stretchLengthInWeeks / repeatBlock.length
  );
  for (i = 0; i < numberOfRepeatBlocks; i++) {
    repeats = repeats.concat(repeatBlock.map(days => [].concat(days)));
  }
  const excessWeeks = repeats.length - stretchLengthInWeeks;
  return repeats.slice(excessWeeks);
};

const getProgramWeeks = (program, length) => {
  const programLength = program.weeks.length;

  if (length === programLength || program.stretchRules === undefined) {
    return program.weeks;
  }

  if (length > programLength) {
    const stretchedStart = getStretchedProgramStart(
      program,
      programLength,
      length
    );
    return [].concat(stretchedStart, program.weeks);
  }

  // TODO: check total length is within program range, or return mimimum
  return []; // TODO: handle shrinking
};

export default {
  getWeekProgram: (targetDate, { targetTime, name, length }) => {
    if (targetDate === null) return null;
    
    const programs = ProgramService.getPrograms(targetTime, name, length);
    if (programs.length === 0) return null;
    const program = programs[0];

    const programWeeks = getProgramWeeks(program, length);
    const firstDate = DateUtils.nextDate(
      targetDate,
      programWeeks.length * -7 + 1
    );
    const maxWeekDistance = programWeeks
      .map(getWeekDistance)
      .reduce((a, b) => Math.max(a, b), 0);

    const maxIntensity = programWeeks
      .map(getIntensity)
      .reduce((max, w) => Math.max(max, w), 0);

    let i = 0;
    const weeks = programWeeks.map(week => {
      const distance = getWeekDistance(week);
      const intensityLevel = getIntensity(week) / maxIntensity;
      const distanceLevel = distance / maxWeekDistance;
      const days = week.map(day => ({
        ...day,
        date: DateUtils.nextDate(firstDate, i++)
      }));
      return { distance, intensityLevel, distanceLevel, days };
    });
    return { weeks };
  }
};
