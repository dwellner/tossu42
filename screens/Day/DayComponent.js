import React from "react";
import propTypes from "prop-types";
import { View, ScrollView, Text } from "react-native";
import DayContentComponent from "./DayContentComponent";
import WeekSummaryComponent from "./WeekSummaryComponent";
import Dimensions from "Dimensions";
import Carousel from "react-native-snap-carousel";
import NoProgram from "../../components/NoProgram";
import { LinearGradient } from "expo";
import Logo from "../../components/Logo";
import Styles from "../../constants/Styles";
import { dateToDateLabel } from "../../utils/Formatters";

export default class DayComponent extends React.Component {
  static propTypes = {
    weekProgram: propTypes.object,
    date: propTypes.string.isRequired,
    targetEvent: propTypes.object.isRequired,
    targetTime: propTypes.number.isRequired,
    maxHr: propTypes.number,
    changeDate: propTypes.func.isRequired
  };

  _renderDay(day) {
    const { maxHr, targetEvent, targetTime } = this.props;

    return (
      <DayContentComponent
        day={day}
        maxHr={maxHr}
        targetEvent={targetEvent}
        targetTime={targetTime}
      />
    );
  }

  render() {
    const { targetEvent, weekProgram, date, changeDate } = this.props;
    if (weekProgram == null) return <NoProgram />;

    const week = weekProgram.weeks.find(
      week => week.days.findIndex(d => d.date === date) >= 0
    );
    if (week == null) return <NoProgram />;
    let days = weekProgram.weeks
      .map(w => w.days)
      .reduce((a, b) => a.concat(b), []);

    let dayIndex = days.findIndex(day => day.date === date);
    const targetEventDate = dateToDateLabel(targetEvent.date);

    return (
      <View
        style={{
          flex: 1,
          alignSelf: "stretch",
          alignItems: "stretch",
          backgroundColor: "#fff"
        }}
      >
        <View
          style={{
            flexDirection: "row",
            justifyContent: "center",
            alignContent: "stretch"
          }}
        >
          <Logo />
          <View
            style={{
              flex: 1,
              alignItems: "center",
              marginLeft: -68
            }}
          >
            <Text style={{ ...Styles.largeContent }}>{targetEvent.name}</Text>
            <Text style={{ ...Styles.lightContent }}>{targetEventDate}</Text>
          </View>
        </View>
        <Text
          style={{
            ...Styles.largeContent,
            marginTop: 0,
            textAlign: "center",
            borderColor: "#eaeaea",
            borderBottomWidth: 1
          }}
        >
          Harjoitusviikko {week.weekNumber}
        </Text>
        <LinearGradient
          style={{ flex: 1, alignSelf: "stretch", alignItems: "stretch" }}
          colors={["#fff", "#eee", "#fff"]}
        >
          <ScrollView>
            <WeekSummaryComponent
              week={week}
              date={date}
              changeDate={changeDate}
            />
            <View style={{ flex: 1 }}>
              <Carousel
                data={days}
                renderItem={({ item }) => this._renderDay(item)}
                sliderWidth={Dimensions.get("window").width}
                firstItem={dayIndex}
                initialNumToRender={Math.max(dayIndex, 25)}
                itemWidth={300}
                enableMomentum={true}
                onBeforeSnapToItem={index => changeDate(days[index].date)}
              />
            </View>
          </ScrollView>
        </LinearGradient>
      </View>
    );
  }
}
