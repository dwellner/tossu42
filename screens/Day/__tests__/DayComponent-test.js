import "react-native";
import React from "react";
import DayComponent from "../DayComponent";
import renderer from "react-test-renderer";
import ProgramModel from "../../../utils/ProgramModel";

it("renders correctly", () => {
  const weekProgram = ProgramModel.getWeekProgram("2018-10-30", {
    targetTime: 240,
    name: "Tossu.com 4 tunnin t√§ysohjelma",
    length: 24
  });

  const week = weekProgram.weeks[0];
  const tree = renderer
    .create(
      <DayComponent
        week={week}
        date={week.days[0].date}
        targetTime={240}
        changeDate={() => {}}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
