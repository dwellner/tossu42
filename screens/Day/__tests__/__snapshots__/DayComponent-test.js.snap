// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ExponentLinearGradient
  colors={
    Array [
      4294967295,
      4293848814,
      4294967295,
    ]
  }
  locations={null}
  style={
    Object {
      "alignItems": "stretch",
      "alignSelf": "stretch",
      "flex": 1,
    }
  }
/>
`;
